### ISSUE TITLE
SASL (external, plain) authentication fails and I don't get registered on the network

### ISSUE THREAD

**Issue by Mikaela** [kaniini/mammon#60]
```
2015-12-20 22:38:16+0200 -- irc.dereferenced.org: *** Looking up your hostname...
2015-12-20 22:38:16+0200 -- irc.dereferenced.org: *** Checking Ident
2015-12-20 22:38:16+0200 -- irc.dereferenced.org: *** Found your hostname: Ace-of-Diamonds.GirlsLikeUs.mikaela.info
2015-12-20 22:38:16+0200 -- irc.dereferenced.org: *** Got Ident response
2015-12-20 22:38:34+0200 -- SASL authentication failed
```

And when I attempt to `/join #mammon,#offtopic` I get

```
2015-12-20 22:41:14+0200 -- You have not registered
2015-12-20 22:41:14+0200 -- You have not registered
```

I am also not seeing registration timeout either.

This started when irc.dereferenced.org was upgraded earlier today.
- WeeChat 1.4-dev (git: v1.3-136-gbd6262d) Copyright (C) 2003-2015, compiled on Dec 15 2015 13:02:07
- ZNC 1.7.x-git-527-e210ade - http://znc.in
  - IPv6: yes, SSL: yes, DNS: threads, charset: yes

**Comment by Mikaela** [comment]
*raw from ZNC in hope you understand it better than I do.

```
2015-12-20 22:47:38+0200 < *raw> YOU -> [znc jump]
2015-12-20 22:47:38+0200 < *raw> YOU -> [WHOIS *raw *raw]
2015-12-20 22:47:39+0200 < *raw> IRC -> [:irc.dereferenced.org NOTICE * :*** Looking up your hostname...]
2015-12-20 22:47:39+0200 < *raw> IRC -> [:irc.dereferenced.org NOTICE * :*** Checking Ident]
2015-12-20 22:47:39+0200 < *raw> IRC -> [:irc.dereferenced.org NOTICE * :*** Found your hostname: Ace-of-Diamonds.GirlsLikeUs.mikaela.info]
2015-12-20 22:47:39+0200 < *raw> IRC -> [:irc.dereferenced.org 451 * :You have not registered]
2015-12-20 22:47:39+0200 < *raw> IRC -> [:irc.dereferenced.org CAP * LS * :userhost-in-names server-time account-notify extended-join away-notify echo-message account-tag cap-notify metadata-notify]
2015-12-20 22:47:39+0200 < *raw> IRC -> [:irc.dereferenced.org CAP * LS sasl]
2015-12-20 22:47:40+0200 < *raw> IRC -> [:irc.dereferenced.org NOTICE Mikaela :*** Got Ident response]
2015-12-20 22:47:40+0200 < *raw> IRC -> [:irc.dereferenced.org CAP Mikaela ACK :account-notify ]
2015-12-20 22:47:41+0200 < *raw> IRC -> [:irc.dereferenced.org CAP Mikaela ACK :away-notify ]
2015-12-20 22:47:42+0200 < *raw> IRC -> [:irc.dereferenced.org CAP Mikaela ACK :extended-join ]
2015-12-20 22:47:43+0200 < *raw> IRC -> [:irc.dereferenced.org CAP Mikaela ACK :sasl ]
2015-12-20 22:47:44+0200 < *raw> IRC -> [:irc.dereferenced.org CAP Mikaela ACK :server-time ]
2015-12-20 22:47:45+0200 < *raw> IRC -> [@time=2015-12-20T20:47:58.000Z :irc.dereferenced.org 904 Mikaela :SASL authentication failed]
2015-12-20 22:48:00+0200 < *raw> YOU -> [PING ::1]
2015-12-20 22:48:00+0200 < *raw> IRC -> [@time=2015-12-20T20:48:14.000Z :irc.dereferenced.org PONG ::1]
2015-12-20 22:48:33+0200 < *raw> YOU -> [JOIN #mammon,#offtopic]
2015-12-20 22:48:33+0200 < *raw> IRC -> [@time=2015-12-20T20:48:47.000Z :irc.dereferenced.org 451 Mikaela :You have not registered]
2015-12-20 22:48:33+0200 < *raw> IRC -> [@time=2015-12-20T20:48:47.000Z :irc.dereferenced.org 451 Mikaela :You have not registered]
2015-12-20 22:49:00+0200 < *raw> YOU -> [PING ::1]
2015-12-20 22:49:00+0200 < *raw> IRC -> [@time=2015-12-20T20:49:14.000Z :irc.dereferenced.org PONG ::1]
```

Why is it replying to my pings if I am not registered? At what point it attempted SASL?

**Comment by Mikaela** [comment]
I think *raw is forgotting some parts and I am able to connect when I unload ZNC's *sasl module.

**Comment by Mikaela** [comment]
And when *SASL is told to only use PLAIN it also results to this issue.

**Comment by DanielOaks** [comment]
I think ZNC's SASL module doesn't like us sending AUTHENTICATE with a source (and would probably also fall over at us sending it with tags as we do when server-time is enabled). [Here](https://github.com/znc/znc/blob/79e979e5cdbcca8d969d1df7f8f370d56e9bdb00/modules/sasl.cpp#L216) is where I think the code is, but I haven't had time to fully track it down and confirm it yet, so I haven't submitted an issue at this time.

In other words, may be a pretty similar issue to https://github.com/hexchat/hexchat/issues/1566 , but I haven't totally confirmed it yet.

**note:** We also send AUTHENTICATE with tags when `server-time` is requested. As mentioned in that HexChat issue, that's something to also keep in mind.

**Comment by DanielOaks** [comment]
Issue reported at https://github.com/znc/znc/issues/1212
